File: src/main/java/ru/darkkeks/vkmirror/vk/object/Message.java
package ru.darkkeks.vkmirror.vk.object;

import com.google.gson.*;

import java.lang.reflect.Type;

public class Message {

    private int messageId;
    private int flags;
    private int peerId;
    private int timestamp;
    private String text;
    private JsonObject attachements;

    private String title;
    private int from;

    public int getMessageId() {
        return messageId;
    }

    public int getFlags() {
        return flags;
    }

    public int getPeerId() {
        return peerId;
    }

    public int getTimestamp() {
        return timestamp;
    }

    public String getText() {
        return text;
    }

    public JsonObject getAttachements() {
        return attachements;
    }

    public String getTitle() {
        return title;
    }

    public int getFrom() {
        return from;
    }

    public static class MessageDeserializer implements JsonDeserializer<Message> {
        @Override
        public Message deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
            JsonArray array = json.getAsJsonArray();

            Message result = new Message();
            result.messageId = array.get(1).getAsInt();
            result.flags = array.get(2).getAsInt();
            if(array.size() > 3) {
                result.peerId = array.get(3).getAsInt();
            }
            if(array.size() > 4) {
                result.timestamp = array.get(4).getAsInt();
                result.text = array.get(5).getAsString();
            }
            if(array.size() > 6) {
                JsonObject additional = array.get(6).getAsJsonObject();
                if(additional.has("title")) {
                    result.title = additional.get("title").getAsString();
                }
                if(additional.has("from")) {
                    result.from = Integer.parseInt(additional.get("from").getAsString());
                }
            }
            if(array.size() > 7) {
                result.attachements = array.get(7).getAsJsonObject();
            }
            if(result.from == 0) {
                result.from = result.peerId;
            }
            return result;
        }
    }

    public static class Flags {
        public static final int UNREAD = 1;
        public static final int OUTBOX = 2;
        public static final int REPLIED = 4;
        public static final int IMPORTANT = 8;
        public static final int CHAT = 16;
        public static final int FRIENDS = 32;
        public static final int SPAM = 64;
        public static final int DELЕTЕD = 128;
        public static final int FIXED = 256;
        public static final int MEDIA = 512;
        public static final int HIDDEN = 65536;
    }
}

File: src/main/java/ru/darkkeks/vkmirror/vk/object/Message.kt
package ru.darkkeks.vkmirror.vk.`object`

import com.google.gson.*
import ru.darkkeks.vkmirror.util.getNullable
import java.lang.reflect.Type

data class Message constructor(
        val messageId: Int,
        val flags: Int,
        val peerId: Int,
        val timestamp: Int,
        val text: String,
        val attachments: JsonObject,
        val attachmentIds: List<Attachment>,
        val additionalInfo: JsonObject,
        val title: String?,
        val from: Int
) {

    class MessageDeserializer : JsonDeserializer<Message> {
        override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Message {
            val array = json.asJsonArray

            val messageId = array[1].asInt
            val flags = array[2].asInt

            val peerId = array.getNullable(3)?.asInt ?: 0

            val timestamp = array.getNullable(4)?.asInt ?: 0
            val text = array.getNullable(5)?.asString ?: ""

            val additional = array.getNullable(6)?.asJsonObject ?: JsonObject()
            val attachments = array.getNullable(7)?.asJsonObject ?: JsonObject()

            val from = additional["from"]?.asString?.toInt() ?: peerId
            val title = additional["title"]?.asString

            val attachmentTypes = mutableListOf<Attachment>()

            val infiniteSeries = generateSequence(1) { it + 1 }
            for (position in infiniteSeries) {
                val prefix = "attach$position"
                if (!attachments.has(prefix)) break

                val id = attachments[prefix].asString
                val type: String = attachments["${prefix}_type"].asString
                val enumType = AttachmentType.valueOf(type)

                attachmentTypes.add(Attachment(id, position, enumType))
            }

            return Message(messageId, flags, peerId, timestamp, text, attachments, attachmentTypes, additional, title, from)
        }
    }

    object Flags {
        const val UNREAD = 1
        const val OUTBOX = 2
        const val REPLIED = 4
        const val IMPORTANT = 8
        const val CHAT = 16
        const val FRIENDS = 32
        const val SPAM = 64
        const val DELETED = 128
        const val FIXED = 256
        const val MEDIA = 512
        const val HIDDEN = 65536
    }

    class Attachment(val id: String, val pos: Int, val type: AttachmentType)

    enum class AttachmentType {
        PHOTO, VIDEO, AUDIO, DOC, WALL, STICKER, LINK, MONEY
    }
}